chain-tx-enclave-next/mls/src/message.rs

```diff
@@ -554,3 +569,93 @@ impl Codec for Sender {
         })
     }
 }
+
+/// Content of commit message and proposals
+pub struct CommitContent {
+    pub sender: LeafSize,
+    pub commit: Commit,
+    pub confirmation: Vec<u8>,
+    pub additions: Vec<Add>,
+    pub updates: Vec<(LeafSize, Update)>,
+    pub removes: Vec<Remove>,
+}
+
+impl CommitContent {
+    /// Verify and extract message contents
+    pub fn new(
+        cs: crate::ciphersuite::CipherSuite,
+        commit: &MLSPlaintext,
+        proposals: &[MLSPlaintext],
+    ) -> Result<Self, ()> {
+        let sender = LeafSize(commit.content.sender.sender);
+        let (commit, confirmation) = match &commit.content.content {
+            ContentType::Commit {
+                commit,
+                confirmation,
+            } => (commit.clone(), confirmation.clone()),
+            _ => {
+                return Err(());
+            }
+        };
+
+        // "Verify that the path value is populated if either of the updates or removes vectors has length greater than zero
+        // all of the updates, removes, and adds vectors are empty."
+        if commit.path.is_none()
+            && ((!commit.updates.is_empty() || !commit.removes.is_empty())
+                || (commit.adds.is_empty()
+                    && commit.updates.is_empty()
+                    && commit.removes.is_empty()))
+        {
+            return Err(());
+        }
+
+        let proposals_ids = proposals
+            .iter()
+            .map(|p| (ProposalId(cs.hash(&p.get_encoding())), p))
+            .collect::<BTreeMap<_, _>>();
+
+        let additions = commit
+            .adds
+            .iter()
+            .map(|proposal_id| {
+                proposals_ids
+                    .get(proposal_id)
+                    .and_then(|add| add.get_add().cloned())
+                    .ok_or(())
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+        let updates = commit
+            .updates
+            .iter()
+            .map(|proposal_id| {
+                proposals_ids
+                    .get(proposal_id)
```

Comment:
depth is too deep, can it be simpilfied?

Reply:
can it be like 'haskell' currying style?

Reply:
yeah, Haskell syntax would be better for this style 😄, but I'm not sure how to do it better in rust.